File: app/src/main/java/com/robotnec/budget/core/mvp/presenter/CalculatorPresenter.java
package com.robotnec.budget.core.mvp.presenter;

import com.robotnec.budget.core.calculator.CalculatorModel;
import com.robotnec.budget.core.calculator.CalculatorModelImpl;
import com.robotnec.budget.core.calculator.Op;
import com.robotnec.budget.core.di.ApplicationComponent;
import com.robotnec.budget.core.exception.InvalidExpressionException;
import com.robotnec.budget.core.mvp.view.CalculatorView;

/**
 * @author zak <zak@swingpulse.com>
 */
public class CalculatorPresenter extends Presenter<CalculatorView> {

    private final CalculatorModel calculatorModel;
    private boolean done;

    public CalculatorPresenter(CalculatorView view) {
        super(view);
        calculatorModel = new CalculatorModelImpl();
        done = false;
    }

    @Override
    public void injectComponent(ApplicationComponent applicationComponent) {
        applicationComponent.inject(this);
    }

    @Override
    public void onViewResume() {
        calculate();
    }

    public void digit(int digit) {
        display(calculatorModel.digit(digit));
    }

    public void dot() {
        display(calculatorModel.dot());
    }

    public void calculate() {
        try {
            String displayText = calculatorModel.calculate();
            double number = Double.parseDouble(displayText);
            if (Double.isInfinite(number) || Double.isNaN(number)) {
                display(displayText);
                getView().displayError();
            } else {
                if (done) {
                    getView().done(Double.parseDouble(displayText));
                } else {
                    display(displayText);
                    done = true;
                    getView().displayDone();
                }
            }
        } catch (InvalidExpressionException e) {
            getView().displayError();
        }
    }

    public void divide() {
        display(calculatorModel.operation(Op.Companion.getDIVIDE()));
    }

    public void multiply() {
        display(calculatorModel.operation(Op.Companion.getMULTIPLY()));
    }

    public void minus() {
        display(calculatorModel.operation(Op.Companion.getMINUS()));
    }

    public void plus() {
        display(calculatorModel.operation(Op.Companion.getPLUS()));
    }

    public void back() {
        display(calculatorModel.back());
    }

    public void clear() {
        display(calculatorModel.clear());
    }

    private void display(String displayText) {
        getView().display(displayText);
        getView().clearError();
        getView().clearDone();
        done = false;
    }
}

File: app/src/main/java/com/robotnec/budget/core/mvp/presenter/CalculatorPresenter.kt
package com.robotnec.budget.core.mvp.presenter

import com.robotnec.budget.refactored.di.ApplicationComponent
import com.robotnec.budget.core.mvp.view.CalculatorView
import org.joda.money.CurrencyUnit
import org.joda.money.Money

/**
 * @author zak zak@swingpulse.com>
 */
class CalculatorPresenter(view: CalculatorView, var amount: Money?)
    : Presenter<CalculatorView>(view) {

    companion object {
        const val MAX_NUMBERS_COUNT: Int = 16
    }

    private val input: MutableList<Int> = mutableListOf()

    override fun injectComponent(applicationComponent: ApplicationComponent) {
        applicationComponent.inject(this)
    }

    override fun onViewResume() {
        view.display(amount.toString())
    }

    fun digit(digit: Int) {
        if (input.size < MAX_NUMBERS_COUNT) {
            input.add(digit)
            view.display(createMoney(input)
                    .also { amount = it }
                    .toString())
        } else {
            view.showMaxCountReached()
        }
    }

    fun done() {
        view.done(createMoney(input))
    }

    fun back() {
        if (input.size > 0) {
            input.removeAt(input.size - 1)
            view.display(createMoney(input)
                    .also { amount = it }
                    .toString())
        } else {
            amount = Money.of(CurrencyUnit.USD, 0.0)
            view.display(amount.toString())
        }
    }

    private fun createMoney(input: MutableList<Int>): Money {
        var value = 0.0
        if (input.size > 0) {
            value = input
                    .joinToString(separator = "")
                    .toDouble() / 100
        }
        return Money.of(CurrencyUnit.USD, value)
    }
}
